src/solver/infeasible-problem-analysis/report.cpp

```diff
@@ -21,107 +21,58 @@
 
 #include <algorithm>
+#include <regex>
 
 #include <antares/logs/logs.h>
-#include "antares/solver/infeasible-problem-analysis/constraint.h"
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-parameter"
-#include "ortools/linear_solver/linear_solver.h"
-#pragma GCC diagnostic pop
-
-using namespace operations_research;
-
-static bool compareSlackSolutions(const Antares::Optimization::Constraint& a,
-                                  const Antares::Optimization::Constraint& b)
-{
-    return a.getSlackValue() > b.getSlackValue();
-}
 
 namespace Antares::Optimization
 {
 InfeasibleProblemReport::InfeasibleProblemReport(
   const std::vector<const MPVariable*>& slackVariables)
 {
-    turnSlackVarsIntoConstraints(slackVariables);
-    sortConstraintsBySlackValue();
-    trimConstraints();
-    sortConstraintsByType();
+    buildConstraintsFromSlackVars(slackVariables);
+    filterConstraintsToOneByType();
 }
 
-void InfeasibleProblemReport::turnSlackVarsIntoConstraints(
+void InfeasibleProblemReport::buildConstraintsFromSlackVars(
   const std::vector<const MPVariable*>& slackVariables)
 {
-    for (const MPVariable* slack: slackVariables)
+    const ConstraintsFactory constraintsFactory;
+    for (const auto& slackVar: slackVariables)
     {
-        constraints_.emplace_back(slack->name(), slack->solution_value());
+        constraints_.push_back(constraintsFactory.create(slackVar->name()));
     }
 }
 
-void InfeasibleProblemReport::sortConstraintsBySlackValue()
-{
-    std::sort(std::begin(constraints_), std::end(constraints_), ::compareSlackSolutions);
-}
-
-void InfeasibleProblemReport::trimConstraints()
-{
-    unsigned int nbConstraints = constraints_.size();
-    constraints_.resize(std::min(nbMaxVariables, nbConstraints));
-}
-
-void InfeasibleProblemReport::sortConstraintsByType()
+void InfeasibleProblemReport::filterConstraintsToOneByType()
 {
-    for (auto& c: constraints_)
+    std::vector<std::string> pickedTypes;
+    for (const auto& c: constraints_)
     {
-        c.extractComponentsFromName();
-        nbConstraintsByType_[c.type()]++;
+        std::string cType = typeid(*c).name();
+        if (std::find(pickedTypes.begin(), pickedTypes.end(), cType) == pickedTypes.end())
+        {
+            uniqueConstraintByType_.push_back(c);
+            pickedTypes.push_back(cType);
+        }
     }
 }
 
 void InfeasibleProblemReport::logSuspiciousConstraints()
 {
-    Antares::logs.error() << "The following constraints are suspicious (first = most suspicious)";
     for (const auto& c: constraints_)
     {
-        Antares::logs.error() << c.prettyPrint();
+        Antares::logs.error() << c->infeasibility();
     }
 }
 
 void InfeasibleProblemReport::logInfeasibilityCauses()
 {
     Antares::logs.error() << "Possible causes of infeasibility:";
-    if (nbConstraintsByType_[ConstraintType::hydro_reservoir_level] > 0)
+    for (const auto& c: uniqueConstraintByType_)
```

Comment:
You can use `std::unique` to remove duplicates. No need to construct this list, just use `constraints_` and remove duplicates (according to type)

See exemple [here](https://en.cppreference.com/w/cpp/algorithm/unique)

Or you can construct a `std::set` as described above.

Reply:
We finally chose to call **std algorithms** in a row instead.